src/main/java/io/cdap/plugin/hive/action/common/HiveConfig.java

```diff
@@ -32,35 +34,147 @@
   public static final String PASSWORD = "password";
   public static final String STATEMENT = "statement";
 
+  private static final String CONNECTION_STRING_PREFIX = "jdbc:hive2://";
+
   @Name(CONNECTION_STRING)
   @Description("JDBC connection string including database name. Please use auth=delegationToken, " +
     "CDAP platform will provide appropriate delegation token while running the pipeline")
   @Macro
-  public String connectionString;
+  private String connectionString;
 
   @Name(USER)
   @Description("User to use to connect to hive metastore database. " +
     "Optional for databases that do not require authentication.")
   @Nullable
   @Macro
-  public String user;
+  private String user;
 
   @Name(PASSWORD)
   @Description("Password to use to connect to hive metastore database. " +
     "Optional for databases that do not require authentication.")
   @Nullable
   @Macro
-  public String password;
+  private String password;
 
   @Name(STATEMENT)
   @Description("Hive command to execute")
   @Macro
-  public String statement;
+  private String statement;
 
   public HiveConfig(String connectionString, String user, String password, String statement) {
     this.connectionString = connectionString;
     this.user = user;
     this.password = password;
     this.statement = statement;
   }
+
+  private HiveConfig(Builder<?> builder) {
+    connectionString = builder.connectionString;
+    user = builder.user;
+    password = builder.password;
+    statement = builder.statement;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static Builder builder(HiveConfig copy) {
+    return builder()
+      .setConnectionString(copy.connectionString)
+      .setUser(copy.user)
+      .setPassword(copy.password)
+      .setStatement(copy.statement);
+  }
+
+  public String getConnectionString() {
+    return connectionString;
+  }
+
+  @Nullable
+  public String getUser() {
+    return user;
+  }
+
+  @Nullable
+  public String getPassword() {
+    return password;
+  }
+
+  public String getStatement() {
+    return statement;
+  }
+
+  public void validate(FailureCollector failureCollector) {
+    if (!containsMacro(CONNECTION_STRING) && !connectionString.startsWith(CONNECTION_STRING_PREFIX)) {
+      failureCollector.addFailure(
+        "Invalid connection string.",
+        "Connection String must comply with format - jdbc:vertica://<VerticaHost>:<portNumber>/<databaseName>")
+        .withConfigProperty(CONNECTION_STRING);
+    }
+
+    if (!containsMacro(USER) && !containsMacro(PASSWORD)) {
+      if (Strings.isNullOrEmpty(user) && !Strings.isNullOrEmpty(password)) {
+        failureCollector.addFailure(
+          "Username is not specified.",
+          "Ensure both username and password are provided.")
```

Comment:
Can't we just remove @Nullable from this properties instead of doing these checks?

Reply:
user and password are the optional fields that's why we can't remove @nullable.